File: matlab/+spine/+systems/system_2_spin_2_singlet.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Author: J.P.G. van Dijk                                                 %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

classdef system_2_spin_2_singlet < spine.systems.system_spin
    
    methods
        
        function obj = system_2_spin_2_singlet()
            obj = obj@spine.systems.system_spin(2);
        end
        
        function index = getIndex(~, states)
            if ((states(1) == spine.systems.system_spin.STATE_0) && (states(2) == spine.systems.system_spin.STATE_0))
                index = 1;
            elseif ((states(1) == spine.systems.system_spin.STATE_0) && (states(2) == spine.systems.system_spin.STATE_1))
                index = 2;
            elseif ((states(1) == spine.systems.system_spin.STATE_1) && (states(2) == spine.systems.system_spin.STATE_0))
                index = 3;
            elseif ((states(1) == spine.systems.system_spin.STATE_1) && (states(2) == spine.systems.system_spin.STATE_1))
                index = 4;
            elseif ((states(1) == spine.systems.system_spin.STATE_N) && (states(2) == spine.systems.system_spin.STATE_S))
                index = 5;
            elseif ((states(1) == spine.systems.system_spin.STATE_S) && (states(2) == spine.systems.system_spin.STATE_N))
                index = 6;
            end
        end
        
        function index = getIndexMeasurement(~, dot, state)
            index = [];
            if (dot == 1)
                if (state == spine.systems.system_spin.STATE_0)
                    index = [1, 2];
                elseif (state == spine.systems.system_spin.STATE_1)
                    index = [3, 4];
                elseif (state == spine.systems.system_spin.STATE_N)
                    index = 5;
                elseif (state == spine.systems.system_spin.STATE_S)
                    index = 6;
                end
            elseif (dot == 2)
                if (state == spine.systems.system_spin.STATE_0)
                    index = [1, 3];
                elseif (state == spine.systems.system_spin.STATE_1)
                    index = [2, 4];
                elseif (state == spine.systems.system_spin.STATE_S)
                    index = 5;
                elseif (state == spine.systems.system_spin.STATE_N)
                    index = 6;
                end
            end
        end
        
        function dimension = getDimension(obj)
            dimension = 6;
        end

        function H = updateHamiltonian(obj)
            
            % Single microwave driveline
            microwave_control = obj.microwave_control(1) + obj.microwave_control(2);
            
            % Common Rabi frequency and charging energy in simple Hamiltonian
            rabi_frequency = obj.rabi_frequency(1);
            charging_energy = obj.charging_energy(1);
            
            % All tunnel rates are the same for a double dot
            tunnel_control = obj.tunnel_control(1, 1);
            
            % Relative detuning
            detuning_control = obj.detuning_control(2) - obj.detuning_control(1);
            
            H = [-(obj.larmor_frequency(1) + obj.larmor_frequency(2)) / 2.0, microwave_control*rabi_frequency, microwave_control*rabi_frequency, 0, 0, 0;
                microwave_control*rabi_frequency, -(obj.larmor_frequency(1) - obj.larmor_frequency(2)) / 2.0, 0, microwave_control*rabi_frequency, tunnel_control, tunnel_control;
                microwave_control*rabi_frequency, 0, +(obj.larmor_frequency(1) - obj.larmor_frequency(2)) / 2.0, microwave_control*rabi_frequency, -tunnel_control, -tunnel_control;
                0, microwave_control*rabi_frequency, microwave_control*rabi_frequency, +(obj.larmor_frequency(1) + obj.larmor_frequency(2)) / 2.0, 0, 0;
                0, tunnel_control, -tunnel_control, 0, charging_energy - detuning_control, 0;
                0, tunnel_control, -tunnel_control, 0, 0, charging_energy + detuning_control];
           
        end
        
    end
    
end

File: cpp/spine/systems/system_2_spin_2_singlet.cpp
/***************************************************************
* Author: J.P.G. van Dijk
***************************************************************/

#include "system_2_spin_2_singlet.h"

using namespace spine;

spine::systems::system_2_spin_2_singlet::system_2_spin_2_singlet()
	: spine::systems::system_spin(2)
{
	
}

spine::systems::system_2_spin_2_singlet::~system_2_spin_2_singlet()
{

}

int spine::systems::system_2_spin_2_singlet::getIndex(unsigned int * states)
{
	if ((states[0] == STATE_0) && (states[1] == STATE_0))
		return 0;
	if ((states[0] == STATE_0) && (states[1] == STATE_1))
		return 1;
	if ((states[0] == STATE_1) && (states[1] == STATE_0))
		return 2;
	if ((states[0] == STATE_1) && (states[1] == STATE_1))
		return 3;
	if ((states[0] == STATE_N) && (states[1] == STATE_S))
		return 4;
	if ((states[0] == STATE_S) && (states[1] == STATE_N))
		return 5;
	return -1;
}

vector<unsigned int> * spine::systems::system_2_spin_2_singlet::getIndexMeasurement(unsigned int dot, unsigned int state)
{
	if (dot == 0) {
		if (state == STATE_0)
			return &measurement_0_0;
		if (state == STATE_1)
			return &measurement_0_1;
		if (state == STATE_N)
			return &measurement_0_N;
		if (state == STATE_S)
			return &measurement_0_S;
	}
	if (dot == 1) {
		if (state == STATE_0)
			return &measurement_1_0;
		if (state == STATE_1)
			return &measurement_1_1;
		if (state == STATE_S)
			return &measurement_1_S;
		if (state == STATE_N)
			return &measurement_1_N;
	}
	return nullptr;
}

unsigned int spine::systems::system_2_spin_2_singlet::getDimension()
{
	return 6;
}

void spine::systems::system_2_spin_2_singlet::updateHamiltonian(realnum * H)
{
	// Single microwave driveline
	realnum microwave_control = this->microwave_control[0] + this->microwave_control[1];

	// Common Rabi frequency and charging energy in simple Hamiltonian
	realnum rabi_frequency = this->rabi_frequency[0];
	realnum charging_energy = this->charging_energy[0];

	// All tunnel rates are the same for a double dot
	realnum tunnel_control = this->tunnel_control[0];

	// Relative detuning
	realnum detuning_control = this->detuning_control[1] - this->detuning_control[0];

	H[0] = -(larmor_frequency[0] + larmor_frequency[1]) / ((realnum) 2.0);
	H[1] = microwave_control * rabi_frequency;
	H[2] = microwave_control * rabi_frequency;
	H[3] = 0;
	H[4] = 0;
	H[5] = 0;
	H[6] = microwave_control * rabi_frequency;
	H[7] = -(larmor_frequency[0] - larmor_frequency[1]) / ((realnum) 2.0);
	H[8] = 0;
	H[9] = microwave_control * rabi_frequency;
	H[10] = tunnel_control;
	H[11] = tunnel_control;
	H[12] = microwave_control * rabi_frequency;
	H[13] = 0;
	H[14] = +(larmor_frequency[0] - larmor_frequency[1]) / ((realnum) 2.0);
	H[15] = microwave_control * rabi_frequency;
	H[16] = -tunnel_control;
	H[17] = -tunnel_control;
	H[18] = 0;
	H[19] = microwave_control * rabi_frequency;
	H[20] = microwave_control * rabi_frequency;
	H[21] = +(larmor_frequency[0] + larmor_frequency[1]) / ((realnum) 2.0);
	H[22] = 0;
	H[23] = 0;
	H[24] = 0;
	H[25] = tunnel_control;
	H[26] = -tunnel_control;
	H[27] = 0;
	H[28] = charging_energy - detuning_control;
	H[29] = 0;
	H[30] = 0;
	H[31] = tunnel_control;
	H[32] = -tunnel_control;
	H[33] = 0;
	H[34] = 0;
	H[35] = charging_energy + detuning_control;
}
